### PR TITLE
Snapshot mana epochs

### PR DESCRIPTION
**Pull request by georgysavva**

### CODE REVIEW

File: plugins/messagelayer/mana_plugin.go [hunk 1]
```diff
@@ -799,11 +799,45 @@ func QueryAllowed() (allowed bool) {
...
+			}
+			return nil
+		})
+
+	}
+	processOutputs(*snapshot.Outputs, snapshot.OutputsMetadata, true /* isCreated */)
+	for i := snapshot.DiffEpochIndex; i < snapshot.FullEpochIndex; i++ {
+		diff, exists := snapshot.EpochDiffs.Get(i)
+		if !exists {
+			continue
```
Comment (karimodm) [inline plugins/messagelayer/mana_plugin.go:835]: This should be a fatal error. There is something wrong with the snapshot if we miss a diff in between fullindex and diffindex.

File: packages/notarization/epoch.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package notarization
```
Comment (karimodm) [inline packages/notarization/epoch.go:1]: I've created a models.go file, these should go in there.

File: packages/snapshot/snapshot.go [hunk 3]
```diff
@@ -18,16 +15,15 @@ import (
 	"github.com/iotaledger/goshimmer/packages/ledger/utxo"
 	"github.com/iotaledger/goshimmer/packages/ledger/vm/devnetvm"
 	"github.com/iotaledger/goshimmer/packages/mana"
+	"github.com/iotaledger/goshimmer/packages/notarization"
 )
 
 type Snapshot struct {
 	LedgerSnapshot *ledger.Snapshot
-	ManaSnapshot   *mana.Snapshot
```
Comment (karimodm) [inline packages/snapshot/snapshot.go:25]: Can we just have the inner struct as this now has only one field?

File: packages/mana/consensusbasevector.go [hunk 4]
```diff
@@ -92,29 +92,29 @@ func txInfoFromPledgeEvent(ev *PledgedEvent) *TxInfo {
...
-				ManaType:      c.Type(),
-				TransactionID: record.TxID,
-			})
-		}
-
-		c.vector[nodeID] = &ConsensusBaseMana{
-			BaseMana1: value,
-		}
-	}
+func (c *ConsensusBaseManaVector) LoadSnapshot(snapshot map[identity.ID]float64) {
```
Comment (karimodm) [inline packages/mana/consensusbasevector.go:103]: Here we just store the base vector right?
Also let's maybe change the function name.

**Review by karimodm — CHANGES_REQUESTED** [review]

File: tools/genesis-snapshot/snapshotcreator/snapshotcreator.go [hunk 5]
```diff
@@ -56,7 +56,6 @@ func CreateSnapshot(genesisTokenAmount uint64, genesisSeedBytes []byte, nodesToP
 
 	return &snapshot.Snapshot{
 		LedgerSnapshot: ledger.NewSnapshot(outputs, outputsMetadata),
```
Comment (daria305) [inline tools/genesis-snapshot/snapshotcreator/snapshotcreator.go:58]: `manaSnapshot := mana.NewSnapshot()` above can also be removed as it is no longer used

File: packages/snapshot/snapshot.go (hunk 3 above)
Comment (georgysavva) [reply packages/snapshot/snapshot.go:25]: I would leave it as is for now. Maybe we will add another field in the future, then we would need to change it back.

File: packages/mana/consensusbasevector.go (hunk 4 above)
Comment (georgysavva) [reply packages/mana/consensusbasevector.go:103]: Yes, we populate the vector with the map data.
Sure, I changed the name.

File: packages/mana/accessbasevector.go [hunk 6]
```diff
@@ -38,25 +38,25 @@ func (a *AccessBaseManaVector) Has(nodeID identity.ID) bool {
...
-		}
-		// trigger events
-		Events.Pledged.Trigger(&PledgedEvent{
-			NodeID:   nodeID,
-			Amount:   record.AccessMana.Value,
-			Time:     record.AccessMana.Timestamp,
-			ManaType: a.Type(),
-		})
-	}
+func (a *AccessBaseManaVector) LoadSnapshot(snapshot map[identity.ID]float64) {
```
[the remark this answers is not in the document]
Comment (georgysavva) [reply packages/mana/accessbasevector.go:49]: Yes, we populate the vector with the map data.
Sure, I changed the name.

**Pull request merged**